### ISSUE TITLE
Rule Request: Enforce private properties in SwiftUI `View`'s

### ISSUE THREAD

**Issue by sindresorhus** [realm/SwiftLint#3173]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New rule request

1. Why should this rule be added? Share links to existing discussion about what
   the community thinks about this.

Apple recommends that state properties in SwiftUI [`View`](https://developer.apple.com/documentation/swiftui/view)'s are private:

> You should only access a state property from inside the view’s body, or from methods called by it. For this reason, declare your state properties as private, to prevent clients of your view from accessing it. - https://developer.apple.com/documentation/swiftui/state

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

**Would trigger:**

```swift
struct ContentView: View {
	@State var foo = ""
}
```

```swift
struct ContentView: View {
	@ObservedObject var foo = Foo()
}
```

```swift
struct ContentView: View {
	@EnvironmentObject var foo
}
```

**Would NOT trigger:**

```swift
struct ContentView: View {
	var foo = ""
}
```

3. Should the rule be configurable, if so what parameters should be configurable?

I would like to see an option to enforce that all properties that are not initializers to be private.

4. Should the rule be opt-in or enabled by default? Why?
   See [README.md](../README.md#opt-in-rules) for guidelines on when to mark a rule as opt-in.

Yes, since it's what Apple recommends.

**Comment by christiankm** [comment]
I created a simple custom rule for this, as a starting point. I will try to make a PR for an official rule.

```
state_private:
    name: "Private SwiftUI State"
    regex: "\\@State\\s*var"
    message: "SwiftUI @State variables should always be marked private."
    severity: warning
```

**Comment by sindresorhus** [comment]
Please keep it open.

**Comment by sindresorhus** [comment]
Please keep it open.

**Comment by mt00chikin** [comment]
FYI, I created https://github.com/realm/SwiftLint/pull/4769 as a starting point to try and address a subset of these concerns. My team was also looking at a custom rule to detect @State properties and require them to be private. It doesn't cover every concern addressed in here, but it could definitely be updated or configured to cover additional cases if desired.
